### PR TITLE
New writer api

### PR DESCRIPTION
**Pull request by mgautierfr**
This is the new api for the writer part of libzim.

The `zim::writer::Creator` has now several methods (instead of `addArticle`) to add the content to a creator :
- `addMetadata` to add a metadata.
- `addRedirection` to add redirection`
- `addItem` to add content (We reuse the name `Item` to be consistent with the reader names).

There is no more `getLayoutPath` has it is never used.
There is a new `getFaviconPath` to implement, return the path of the favicon (scrapper doesn't have to add themselves the `-/favicon` redirection).

The notion of namespace has been removed from the API. However, this is still present, scrapper must return path containing the namespace ("A/foo.html"). Namespace will be totally removed in next PR.

On the `Item` (new name for `Article`) side :
- Some methods stay identical : `getPath`, `getTitle` and `getMimetype`.
- `getContentProvider` : return a object in charge of providing the item content to zim creator.
- `getIndexData` : return a object representing the data to index (may be different than the data to store).
- `getHints` : to give hints to the creator about the item.

`ContentProvider` is a "small" object in charge to pass the data to the creator.
Two basic `ContentProvider` are provided : `StringProvider` (using content stored in a `std::string`) and `FileProvider` (using content stored in a file).

`IndexData` is used to pass content to index. If not reimplemented, the default implementation does the same as now, based of the content provided by the contentProvider.

Some functional changes with this change :

- No layout page.
- creator automatically detect if the item must be compressed (based on the mimetype). It is possible to force the compression (or deactivate it) using a hint.
- No more `targetLink` or `delete` dirent.
- Item are "title indexed" if they are html and have a title. (No base on previous `shouldIndex`.
- Redirection are "title indexed" if they have a title.
- MainPage entry will not be directly the mainpage but a redirection to the mainpage.
- "Big memory test" doesn't use 8GiB but 4GiB

This is not totally finished :
- No documentation
- User code has not be ported to the new api.

But this is already ready for review (review commit by commit can help).

### CODE REVIEW

File: include/zim/writer/item.h [hunk 1]
```diff
@@ -45,13 +51,15 @@ namespace zim
     class Item
     {
       public:
+        using Hints = std::map<HintKeys, uint64_t>;
```
Comment (veloman-yunkan) [inline include/zim/writer/item.h:61]: What is the meaning of `Hints::value_type`?

File: include/zim/writer/item.h [hunk 2]
```diff
@@ -61,6 +61,55 @@ namespace zim
         virtual ~Item() = default;
     };
 
+    class BasicItem : public Item
+    {
+      public:
+        BasicItem(const std::string& path, const std::string& mimetype, const std::string title)
```
Comment (veloman-yunkan) [inline include/zim/writer/item.h:141]: Why this discrimination of `title`?

File: include/zim/writer/item.h [hunk 3]
```diff
@@ -61,6 +61,55 @@ namespace zim
...
+
+        std::string getPath() const { return path; }
+        std::string getTitle() const { return title; }
+        std::string getMimeType() const { return mimetype; }
+
+      protected:
+        std::string path;
+        std::string mimetype;
+        std::string title;
+    };
```
Comment (veloman-yunkan) [inline include/zim/writer/item.h]: I would change this into `struct ItemDescriptor` that is NOT derived from `Item`, and then derive both `StringItem` and `FileItem` from both `Item` and `ItemDescriptor`. The constructor signatures will then be shortened to

```
StringItem(const ItemDescriptor& itemDesc, const std::string& content);
FileItem(const ItemDescriptor& itemDesc, const std::string& filePath);
```

File: include/zim/writer/creator.h [hunk 4]
```diff
@@ -169,12 +185,12 @@ namespace zim
...
       private:
         std::unique_ptr<CreatorData> data;
-        bool verbose;
-        const CompressionType compression;
-        bool withIndex = false;
-        size_t minChunkSize = 1024-64;
-        std::string indexingLanguage;
-        unsigned nbWorkerThreads = 4;
+        bool m_verbose = false;
+        CompressionType m_compression = zimcompLzma;
```
Comment (veloman-yunkan) [inline include/zim/writer/creator.h:183]: Default compression is now zstd

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]
I have too little familiarity with the writer side of libzim, so this was at least a good opportunity for me to get to know it better. Please take my review comments with a corresponding grain of salt.

File: include/zim/writer/item.h (hunk 1 above)
Comment (mgautierfr) [reply include/zim/writer/item.h:61]: I don't see a use case where the value would be a free value.
All use case I see is about a fixed set of values. And a integer is good enough to identify a set of enumerated values.

Using a string would need a string parsing that could be be time consuming for nothing.
If we need to pass free value, not interpreted by libzim we should probably pass them using other API than a hint.

File: src/writer/workers.cpp [hunk 5]
```diff
@@ -93,50 +78,46 @@ namespace zim
...
-      auto content = zim::removeAccents(htmlParser.dump);
-
-      document.add_value(0, title);
+      document.set_data(p_item->getPath());
+      document.add_value(0, p_item->getTitle());
 
       std::stringstream countWordStringStream;
-      countWordStringStream << countWords(htmlParser.dump);
+      countWordStringStream << indexData->getWordCount();
       document.add_value(1, countWordStringStream.str());
```
[the remark this answers is not in the document]
Comment (mgautierfr) [reply src/writer/workers.cpp:96]: Yes, but android compiler do not implement `to_string` (https://stackoverflow.com/questions/22774009/android-ndk-stdto-string-support)

File: src/writer/creator.cpp [hunk 6]
```diff
@@ -183,8 +210,9 @@ namespace zim
         data->indexer->indexingPostlude();
         microsleep(100);
         auto item = data->indexer->getMetaItem();
-        auto dirent = data->createDirentFromItem(item);
-        data->addDirent(dirent, item);
+        auto dirent = data->createItemDirent(item);
+        data->addDirent(dirent);
+        data->addItemData(dirent, item);
```
[the remark this answers is not in the document]
Comment (mgautierfr) [reply src/writer/creator.cpp]: This is now "used" only once.

**Review by veloman-yunkan — COMMENTED** [review]
I see that you addressed the previous review comments while rebasing. That complicates doing an incremental review. I suggest that for large PRs changes addressing review comments are made as separate commits. They can be squashed during a final rebase right before merging.

So I reviewed only the new commits and came up only with a few spell-checker findings.

**Comment by mgautierfr** [comment]
Typos are fixed.
I've create a different commit that will be squash correctly just before merging once PR is validated.

**Review by veloman-yunkan — APPROVED** [review]

**Pull request merged**